### ISSUE TITLE
[new-feature-request] Use latest version as schema reference

### ISSUE THREAD

**Issue by FabioBatSilva** [confluentinc/terraform-provider-confluent#207]
I have a  resource that creates multiple schemas.
The list of schemas is large so having a way to dynamically define is important.

I'd like to have the latest version of a schema always pointing to the latest schema references.
Being able to use `version = "latest"` would make this possible.

```terraform
resource "confluent_schema" "schemas" {
  for_each = var.schemas

  schema_registry_cluster {
    id = data.confluent_schema_registry_cluster.registry.id
  }

  recreate_on_update = false
  schema             = each.value.schema
  format             = each.value.format
  subject_name       = each.value.subject_name
  rest_endpoint      = data.confluent_schema_registry_cluster.registry.rest_endpoint

  credentials {
    key    = var.confluent_cloud_registry_api_key
    secret = var.confluent_cloud_registry_api_secret
  }

  lifecycle {
    prevent_destroy = true
  }

  dynamic "schema_reference" {
    for_each = each.value.references
    content {
      version      = "latest"
      name         = schema_reference.value.name
      subject_name = schema_reference.value.subject_name
    }
  }
}
```

> **Note**
> This is not possible via terraform lookups but would be the equivalent to : 
> ` version = lookup(confluent_schema.schemas, schema_reference.value.subject_name).version`

**Comment by linouk23** [comment]
Thanks for the note @FabioBatSilva!

What do you think about introducing `confluent_latest_schema` data source that accepts `subject_name` as a input?

**Comment by FabioBatSilva** [comment]
I was thinking about that option too..

I think it would have to be something `confluent_schemas`,
that would return a empty list if the schema is not defined yet..

```terraform
resource "confluent_schemas" "schemas" {
  for_each = var.schemas

  subject_name = each.value.subject_name
  latest_only = true
}
```

Then:
```terraform
dynamic "schema_reference" {
    for_each = toset([
      for s in each.value.references : s if length(lookup(data.confluent_schemas.schemas, s.subject_name, [])) > 0
    ])

    content {
      name         = schema_reference.value.name
      subject_name = schema_reference.value.subject_name
      version      = data.confluent_schemas.schemas[schema_reference.value.name][0].version
    }
  }
```

The issue with this is that on the first run.  where the reference was not created yet
The resource that has the references would be created without then.. and on the next run it would get updated to add the references..

The ` latest or -1` thing sound promising..
Was able to create some records using it `-1`..  :astonished: 

currently testing out some scenarios..

**Comment by FabioBatSilva** [comment]
What i found so far :

`#1`

Running a plan with the exact same values is detected as a change, bc if the `-1` is different that what is currently being used as referenced number.

```
  ~ resource "confluent_schema" "schemas" {
        id                 = "xxxx-xxxxx/examples.terraform.v1alpha.other_record-value/latest"
        # (8 unchanged attributes hidden)
      + schema_reference {
          + name         = "examples/terraform/v1alpha/record.proto"
          + subject_name = "examples.terraform.v1alpha.record-value"
          + version      = -1
        }
      - schema_reference {
          - name         = "examples/terraform/v1alpha/record.proto" -> null
          - subject_name = "examples.terraform.v1alpha.record-value" -> null
          - version      = 3 -> null
        }

        # (2 unchanged blocks hidden)
    }
```

`#2`
Looks like changes to a reference don't create a new version.

For instance at the beginning current latest version of `other_record-value` is `1`
`other_record-value` is referencing version `2` of `record-value`

Now  `record-value` gets updated and version `3` is created.
The current latest version of `other_record-value` is still `1` but now it is  referencing `3` of `record-value`


`#1` seems like something that could be fixed by handeling the `-1` version..
`#2` i'm not sure if works that way by design on the api level.. :thinking:

**Comment by linouk23** [comment]
Oh interesting, yeah seems like adding `latest_only` attribute to schema data source might be the easiest option.

**Comment by FabioBatSilva** [comment]
Opened another issue for `confluent_schemas` : https://github.com/confluentinc/terraform-provider-confluent/issues/211
 
 I think might be useful independently of this use case.

**Comment by linouk23** [comment]
@FabioBatSilva did you try [this](https://registry.terraform.io/providers/confluentinc/confluent/latest/docs/resources/confluent_schema#default-option-1-manage-the-latest-schema-version-only-the-resource-instance-always-points-to-the-latest-schema-version-by-supporting-in-place-updates) already?

**Comment by FabioBatSilva** [comment]
> @FabioBatSilva did you try [this](https://registry.terraform.io/providers/confluentinc/confluent/latest/docs/resources/confluent_schema#default-option-1-manage-the-latest-schema-version-only-the-resource-instance-always-points-to-the-latest-schema-version-by-supporting-in-place-updates) already?

The schema evolution / `recreate_on_update` flag ?
Yes, that works for a single resource..
the problem we are having is using that latest version in a reference.. as the scenario included in the description.

**Comment by linouk23** [comment]
@FabioBatSilva did #211 solved this problem or you believe we might want to find a different solution here?

```
data "confluent_schemas" "main" {
  filter {
    subject_prefix = "examples.record"
    latest_only    = true
    deleted         = false
  }
}
```

**Comment by FabioBatSilva** [comment]
Using `confluent_schemas`  is a workaround IMO..
Given this scenario : https://github.com/confluentinc/terraform-provider-confluent/issues/207#issuecomment-1439062635

I think ideally this would be transparently handled by the provider or api.

**Comment by linouk23** [comment]
Sure thing, so I took another look at [your comment](https://github.com/confluentinc/terraform-provider-confluent/issues/207#issuecomment-1439062635) and here're my thinking:

Since schemas could be managed by multiple users and multiple clients (CLI, TF, UI etc) it might be pretty risky to use `version      = -1` directly in TF configuration:
```
schema_reference {
          + name         = "examples/terraform/v1alpha/record.proto"
          + subject_name = "examples.terraform.v1alpha.record-value"
          + version      = -1
        }
```

More specifically, if someone creates a schema `foo` using:
```
schema_reference {
          + name         = "examples/terraform/v1alpha/record.proto"
          + subject_name = "examples.terraform.v1alpha.record-value"
          + version      = -1
        }
```
and then someone updates `v1alpha.record-value` and `v1alpha/record.proto` via UI it might cause TF drift.

All that to say, that we might want to rewrite it as:
```
data "confluent_schema" "latest_purchase" {
  subject_name      = confluent_schema.purchase.subject_name
  latest = true
}

resource "confluent_schema" "customer-event" {
  subject_name = "customer-event-value"
  format = "AVRO"
  schema = file("./schemas/avro/customer_event.avsc")

  schema_reference {
    name         = "Purchase"
    subject_name = confluent_schema.purchase.subject_name
    version      = data.confluent_schema.purchase.version
  }
}
```

so we'd see drift for a data source first and then we'll see an explicit update like `version = 7 -> 8` for `confluent_schema.customer-event`. In other words, we might want to add `confluent_schema` data source (on top of https://registry.terraform.io/providers/confluentinc/confluent/latest/docs/data-sources/confluent_schemas).


How do you like this idea?

**Comment by FabioBatSilva** [comment]
Thanks @linouk23 ,

I think that has the same issues as i mentioned using `confluent_schemas` as we are using today..

The version in `data.confluent_schema.purchase.version` is available only after the resources gets created / updated.
So the version used in `schema_reference` is behind every time there is a change in the data resource that is being looked up..

Creating it as 2 separated resource then using `confluent_schema.purchase.version` would work 
```terraform
resource "confluent_schema" "purchase" {
 # ...
}

resource "confluent_schema" "customer-event" {
  # ...
  schema_reference {
    name         = "Purchase"
    subject_name = confluent_schema.purchase.subject_name
    version      = confluent_schema.purchase.version
  }
}
```

But that doesn't scale if you have to manage many topics..
it is impractical to define all the topics one by one..

**Comment by FabioBatSilva** [comment]
Another way I've seen this implemented in TF is to have a separated resource.

Something like this:

```terraform
resource "confluent_schema" "schemas" {
  for_each = var.schemas

  format             = each.value.format
  subject_name       = each.value.subject_name
  
  # ...
}

resource "confluent_schema_reference" "references" {
  for_each = flatten([
    for schema in var.schemas : schema.references
  ])

  subject_id       = confluent_schema.schemas[each.value.subject_name].id

  name         = confluent_schema.schemas[each.value.subject_name].name
  version      = confluent_schema.schemas[each.value.subject_name].version
  subject_name = confluent_schema.schemas[each.value.subject_name].subject_name
}
```

But i understand that might be tricky to implement given the way the api is defined..

**Comment by FabioBatSilva** [comment]
Yes, `-1` / `latest` as is implemented by the api is also not great..

I think the best would be something like `confluent_schema_reference`
or way for the provider to internally do the equivalent to `version = lookup(confluent_schema.schemas, schema_reference.value.subject_name).version`

But not sure what is actuality possible..

**Comment by noahseger** [comment]
Hey @linouk23 just wondering if you have any direction on the references yet or if it's somewhere on the roadmap?

We're kind of at the point where we're thinking of pre-registering all the latest schemas before applying, because any new schema with references requires 2+ runs to get the state flush, but don't want to start down that road if a fix is incoming.

**Comment by linouk23** [comment]
@noahseger thanks for asking! 

As of right now we haven't figure out a smart way to fix it.

**Comment by noahseger** [comment]
Hey @linouk23 just a heads up that we are going in the direction of using Pulumi for this use case so we can declare the dependencies programmatically.

At this point I think it might be worth a docs solution to close this issue:

1. a "danger ahead" warning for other folks dynamically managing schemas + references with Gitops to start with Pulumi (or whatever else you would recommend)
2. a working example of creating 2 schemas + a reference in a single Terraform run

What do you think?
